File: Backend/app/pw_library/python_methods/lu_factorization_with_pivoting.py
import numpy as np


def lu_factorization_with_pivoting(A):
    """
    Performs LU factorization with partial pivoting on matrix A.
    Returns matrices P, L, U such that PA = LU.
    """
    n = len(A)
    L = np.eye(n)
    U = A.copy()
    P = np.eye(n)

    for i in range(n):
        # Find the pivot row
        pivot_row = np.argmax(np.abs(U[i:, i])) + i

        # Swap rows in U and update P and L accordingly
        if pivot_row != i:
            U[[i, pivot_row]] = U[[pivot_row, i]]
            P[[i, pivot_row]] = P[[pivot_row, i]]
            if i > 0:
                L[[i, pivot_row], :i] = L[[pivot_row, i], :i]

        # Eliminate entries below the pivot
        for j in range(i + 1, n):
            if U[i, i] == 0:
                raise ValueError("Zero pivot encountered, matrix may be singular.")

            # Calculate the multiplier and update L
            L[j, i] = U[j, i] / U[i, i]

            # Perform elimination
            U[j, i:] = U[j, i:] - L[j, i] * U[i, i:]

    return P, L, U


def forward_substitution(L, b):
    """
    Solves the equation Ly = b for y, where L is a lower triangular matrix.
    """
    n = len(b)
    y = np.zeros_like(b, dtype=float)

    for i in range(n):
        y[i] = (b[i] - np.dot(L[i, :i], y[:i])) / L[i, i]

    return y


def backward_substitution(U, y):
    """
    Solves the equation Ux = y for x, where U is an upper triangular matrix.
    """
    n = len(y)
    x = np.zeros_like(y, dtype=float)

    for i in range(n - 1, -1, -1):
        x[i] = (y[i] - np.dot(U[i, i + 1:], x[i + 1:])) / U[i, i]

    return x


def solve_lu_with_pivoting(A, b):
    """
    Solves the linear system Ax = b using LU factorization with partial pivoting.
    """
    # Perform LU factorization with pivoting
    P, L, U = lu_factorization_with_pivoting(A)

    # Adjust b according to the permutation matrix P
    Pb = np.dot(P, b)

    # Solve Ly = Pb for y
    y = forward_substitution(L, Pb)

    # Solve Ux = y for x
    x = backward_substitution(U, y)

    return x, P, L, U

File: Backend/app/pw_library/rust_methods/lu_factorization_with_pivoting.rs
use nalgebra::{DMatrix, DVector};

// LU Factorization with Partial Pivoting
fn lu_factorization_with_pivoting(a: &DMatrix<f64>) -> Result<(DMatrix<f64>, DMatrix<f64>, DMatrix<f64>), &'static str> {
    let n = a.nrows();
    let mut l = DMatrix::identity(n, n);
    let mut u = a.clone();
    let mut p = DMatrix::identity(n, n);

    for i in 0..n {
        // Find the pivot row
        let pivot_row = (i..n).max_by(|&x, &y| u[(x, i)].abs().partial_cmp(&u[(y, i)].abs()).unwrap()).unwrap();

        // Swap rows in U and update P and L accordingly
        if pivot_row != i {
            u.swap_rows(i, pivot_row);
            p.swap_rows(i, pivot_row);
            if i > 0 {
                l.swap_rows_generic(i.into(), pivot_row.into());
            }
        }

        // Eliminate entries below the pivot
        for j in (i + 1)..n {
            if u[(i, i)] == 0.0 {
                return Err("Zero pivot encountered, matrix may be singular.");
            }

            // Calculate the multiplier and update L
            l[(j, i)] = u[(j, i)] / u[(i, i)];

            // Perform elimination
            for k in i..n {
                u[(j, k)] -= l[(j, i)] * u[(i, k)];
            }
        }
    }

    Ok((p, l, u))
}

// Forward substitution for Ly = b
fn forward_substitution(l: &DMatrix<f64>, b: &DVector<f64>) -> DVector<f64> {
    let n = b.len();
    let mut y = DVector::zeros(n);

    for i in 0..n {
        let sum = l.row(i).columns(0, i).dot(&y.rows(0, i));
        y[i] = (b[i] - sum) / l[(i, i)];
    }

    y
}

// Backward substitution for Ux = y
fn backward_substitution(u: &DMatrix<f64>, y: &DVector<f64>) -> DVector<f64> {
    let n = y.len();
    let mut x = DVector::zeros(n);

    for i in (0..n).rev() {
        let sum = u.row(i).columns(i + 1, n - i - 1).dot(&x.rows(i + 1, n - i - 1));
        x[i] = (y[i] - sum) / u[(i, i)];
    }

    x
}

// Solving Ax = b using LU factorization with partial pivoting
fn solve_lu_with_pivoting(a: &DMatrix<f64>, b: &DVector<f64>) -> Result<(DVector<f64>, DMatrix<f64>, DMatrix<f64>, DMatrix<f64>), &'static str> {
    // Perform LU factorization with pivoting
    let (p, l, u) = lu_factorization_with_pivoting(a)?;

    // Adjust b according to the permutation matrix P
    let pb = &p * b;

    // Solve Ly = Pb for y
    let y = forward_substitution(&l, &pb);

    // Solve Ux = y for x
    let x = backward_substitution(&u, &y);

    Ok((x, p, l, u))
}
